x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DeprecationRoleDescriptorPreprocessor.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.authz.store;
+
+import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
+
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.indices.alias.get.GetAliasesAction;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.metadata.AliasMetaData;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.common.logging.DeprecationLogger;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor.IndicesPrivileges;
+import org.elasticsearch.xpack.core.security.authz.permission.IndicesPermission;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Predicate;
+
+import static org.elasticsearch.xpack.core.ClientHelper.SECURITY_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+public final class DeprecationRoleDescriptorPreprocessor implements BiConsumer<Set<RoleDescriptor>, ActionListener<Set<RoleDescriptor>>> {
+
+    private final Logger logger;
+    private final DeprecationLogger deprecationLogger;
+    private final Client client;
+
+    public DeprecationRoleDescriptorPreprocessor(Client client, Logger logger) {
+        this.logger = logger;
+        this.deprecationLogger = new DeprecationLogger(logger);
+        this.client = client;
+    }
+
+    @Override
+    public void accept(Set<RoleDescriptor> effectiveRoleDescriptors, ActionListener<Set<RoleDescriptor>> roleBuilder) {
+        executeAsyncWithOrigin(client, SECURITY_ORIGIN, GetAliasesAction.INSTANCE,
+                client.admin().indices().prepareGetAliases("_all").request(), ActionListener.wrap(getAliasesResponse -> {
+                    final ImmutableOpenMap<String, List<AliasMetaData>> allAliasesMap = getAliasesResponse.getAliases();
+                    // iterate over all effective role descriptors and their indices privileges
+                    for (final RoleDescriptor roleDescriptor : effectiveRoleDescriptors) {
+                        for (final IndicesPrivileges indicesPrivilege : roleDescriptor.getIndicesPrivileges()) {
```

Comment:
Yeah, it was a conscious decision.

The Role Descriptors could overlap with others, as you described, or as a simpler example, it could overlap with _superuser_, and after the overlap the deprecated behavior would be nullified. But the fact of the matter is that such a Role Descriptor _exists_, and it could latter be mapped individually to other users where it _might_ exhibit the deprecated behavior. We could instead verify that indeed the effective Role of the user manifests the behavior, but in this case the log message would be less useful because there could be no indication about which Role Descriptor(s) have to be modified (it is impossible to separate). Therefore I decided to go with the actionable alternative.

Reply:
When first answering this question I had in mind index privileges across multiple Roles.
But, I now realize, the question might have been about index privileges in the same Role Descriptor.

You are right Jay, I have refactored it twice 🙄 to make it work as expected.